daffodil-core/src/main/scala/org/apache/daffodil/grammar/primitives/ChoiceCombinator.scala

```diff
@@ -199,8 +206,13 @@ case class ChoiceCombinator(ch: ChoiceTermBase, alternatives: Seq[Gram])
       val mapValues = eventUnparserMap.map { case (k, v) => v }.toSeq.filterNot(_.isEmpty)
       if (mapValues.isEmpty)
         new NadaUnparser(null)
-      else
-        new ChoiceCombinatorUnparser(ch.modelGroupRuntimeData, eventUnparserMap)
+      else {
+        val cup = new ChoiceCombinatorUnparser(ch.modelGroupRuntimeData, eventUnparserMap)
+        ch.choiceLengthKind match {
+          case ChoiceLengthKind.Implicit => cup
+          case ChoiceLengthKind.Explicit => new ChoiceLengthUnparser(cup, ch.choiceRuntimeData, choiceLengthInBits)
```

Comment:
I think this won't work. ChoiceLengthUnparser has slots in it that are side-effected at runtime to capture length. That won't work if the unparser is being shared across threads.

Reply:
You need the choice combinator unparser to dynamically allocate this ChoiceLengthUnparser at runtime. Not create it here in this grammar-tree.